[Concurrency] Handle cases where a property initializer is subsumed by another property for `IsolatedDefaultValues`.

hborla opened the pull request:
When determining whether compiler synthesized initializers should have `nonisolated` applied, skip property initializers that are subsumed, e.g. by an init accessor or a backing property wrapper initializer, and always consider the subsuming initializer. This fixes an issue where the compiler would apply `nonisolated` to memberwise and default initializers of types that use property wrappers whose backing property wrapper initializer is global actor isolated, but the wrapped value initializer is not, e.g.

```swift
@MainActor
@propertyWrapper struct RequiresMain<Value>  {
  var wrappedValue: Value

  init(wrappedValue: Value) {
    self.wrappedValue = wrappedValue
  }
}

// `S.init()` should be isolated to `@MainActor` because 
// `RequiresMain.init(wrappedValue:)` is isolated to `@MainActor`
struct S {
  @RequiresMain private var x = 10
}
```

This change also lessens the source break of SE-0411 by still emitting member initializers in implicit constructors when the initializer violates actor isolation to preserve the behavior of existing code when concurrency diagnostics are downgraded to warnings in Swift 5 mode.

Resolves rdar://121282537

hborla committed:
[Concurrency] Handle cases where a property initializer is subsumed by another
property for IsolatedDefaultValues.

For property wrappers and init accesors, skip property initializers that are
subsumed, e.g. by an init accessor or a backing property wrapper initializer,
and always consider the subsuming initializer to determine whether compiler
synthesized initializers should have `nonisolated` applied.

This change also lessens the source break of SE-0411 by still emitting
member initializers in implicit constructors when the initializer violates
actor isolation to preserve the behavior of existing code when concurrency
diagnostics are downgraded to warnings in Swift 5 mode.

hborla commented:
@swift-ci please smoke test

hborla commented:
@swift-ci please test source compatibility

DougGregor commented on lib/Sema/CodeSynthesis.cpp:
```diff
@@ -351,34 +351,59 @@ static ConstructorDecl *createImplicitConstructor(NominalTypeDecl *decl,
+
+      auto *var = pbd->getSingleVar();
+      if (!var) {
+        shouldAddNonisolated = false;
```
I don't understand why a pattern without a single variable should suppress `nonisolated`.

hborla replied:
Oh, it shouldn't. I skipped them because I thought they were banned, but we only ban stored properties without a single var for struct `let`s. I'll fix this, thank you!

hborla replied:
Fixed in https://github.com/apple/swift/pull/71033/commits/cc1fd1ce8fb13518f8d7f42e37c2e995d03c3b23

hborla committed:
[Concurrency] Don't skip pattern bindings with multiple vars when inferring the
isolation of implicit initializers.

hborla commented:
Those source compatibility failures are real. This fails under `-strict-concurrency=complete`

```swift
struct CError: Error, RawRepresentable {
  var rawValue: CInt
}
```

Probably because I'm calling `getAllMembers()` which invokes the conformance synthesis earlier. Instead, I can iterate over the stored properties and the init accessor properties, because those are what can participate in the implicit initializers.

hborla committed:
[Concurrency] Don't call `getAllMembers()` while inferring actor isolation for
implicit initializers.

This messes with conformance synthesis for `RawRepresentable` and friends because
it invokes conformance synthesis too early.

hborla commented:
@swift-ci please smoke test

hborla commented:
@swift-ci please test source compatibility
